--- a/src/amd/vulkan/radv_device.c
+++ b/src/amd/vulkan/radv_device.c
@@ -51,7 +51,7 @@
 #include "util/timespec.h"
 #include "util/u_atomic.h"
 #include "winsys/null/radv_null_winsys_public.h"
-#include "ac_llvm_util.h"
+
 #include "git_sha1.h"
 #include "sid.h"
 #include "vk_format.h"
@@ -108,8 +108,7 @@
    memset(uuid, 0, VK_UUID_SIZE);
    _mesa_sha1_init(&ctx);
 
-   if (!disk_cache_get_function_identifier(radv_device_get_cache_uuid, &ctx) ||
-       !disk_cache_get_function_identifier(LLVMInitializeAMDGPUTargetInfo, &ctx))
+   if (!disk_cache_get_function_identifier(radv_device_get_cache_uuid, &ctx))
       return -1;
 
    _mesa_sha1_update(&ctx, &family, sizeof(family));
@@ -293,7 +292,7 @@
       return "ACO";
    }
 
-   return "LLVM " MESA_LLVM_VERSION_STRING;
+   return "ACO";
 }
 
 int
@@ -622,7 +621,7 @@
    device->local_fd = fd;
    device->ws->query_info(device->ws, &device->rad_info);
 
-   device->use_llvm = instance->debug_flags & RADV_DEBUG_LLVM;
+   device->use_llvm = 0;
 
    snprintf(device->name, sizeof(device->name), "AMD RADV %s (%s)", device->rad_info.name,
             radv_get_compiler_string(device));
