From cd8dee90a59f1d20b827a6d6c8db10d30c159ab8 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <z.figura12@gmail.com>
Date: Fri, 8 Jun 2018 15:41:01 -0500
Subject: [PATCH 09/83] ntdll: Implement NtSetEvent().

Signed-off-by: Zebediah Figura <z.figura12@gmail.com>
---
 dlls/ntdll/sync.c  |  3 +++
 3 files changed, 19 insertions(+)


diff --git a/dlls/ntdll/sync.c b/dlls/ntdll/sync.c
index 84abb0d205..c8047bfa3e 100644
--- a/dlls/ntdll/unix/sync.c
+++ b/dlls/ntdll/unix/sync.c
@@ -487,7 +487,11 @@ NTSTATUS WINAPI NtSetEvent( HANDLE handle, LONG *prev_state )
 NTSTATUS WINAPI NtSetEvent( HANDLE handle, LONG *prev_state )
 {
     unsigned int ret;
+
+    if (do_fsync())
+        return fsync_set_event( handle, prev_state );
+

     SERVER_START_REQ( event_op )
     {
         req->handle = wine_server_obj_handle( handle );
--
2.19.1

