--- a/CMakeLists.txt	2020-09-17 20:08:11.387064584 +0400
+++ b/CMakeLists.txt	2020-09-17 20:08:15.727064567 +0400
@@ -1178,12 +1178,6 @@
     check_include_file("libudev.h" HAVE_LIBUDEV_H)
 
     if(PKG_CONFIG_FOUND)
-      pkg_search_module(DBUS dbus-1 dbus)
-      if(DBUS_FOUND)
-        set(HAVE_DBUS_DBUS_H TRUE)
-        include_directories(${DBUS_INCLUDE_DIRS})
-        list(APPEND EXTRA_LIBS ${DBUS_LIBRARIES})
-      endif()
 
       pkg_search_module(IBUS ibus-1.0 ibus)
       if(IBUS_FOUND)
