--- a/include/wine/vulkan_driver.h
--- b/include/wine/vulkan_driver.h
@@ -44,6 +47,8 @@

     /* winevulkan specific functions */
     VkSurfaceKHR (*p_wine_get_host_surface)(VkSurfaceKHR);
+
+    VkBool32 (*query_fsr)(VkSurfaceKHR surface, VkExtent2D *real_sz, VkExtent2D *user_sz, VkRect2D *dst_blit, VkFilter *filter, BOOL *fsr, float *sharpness);
 };

 static inline void *get_vulkan_driver_device_proc_addr(
