--- a/include/wine/vulkan_driver.h
--- b/include/wine/vulkan_driver.h
@@ -44,6 +47,8 @@

     /* winevulkan specific functions */
     VkSurfaceKHR (*p_wine_get_native_surface)(VkSurfaceKHR);
+
+    VkBool32 (*query_fs_hack)(VkSurfaceKHR surface, VkExtent2D *real_sz, VkExtent2D *user_sz, VkRect2D *dst_blit, VkFilter *filter, BOOL *fsr, float *sharpness);
 };

 extern const struct vulkan_funcs * CDECL __wine_get_vulkan_driver(UINT version);
