--- a/dlls/winevulkan/vulkan_thunks.c
+++ b/dlls/winevulkan/vulkan_thunks.c
@@ -44933,7 +44933,8 @@

     init_conversion_context(ctx);
     convert_VkPresentInfoKHR_win64_to_driver(ctx, params->pPresentInfo, &pPresentInfo_host);
-    params->result = wine_queue_from_handle(params->queue)->device->funcs.p_vkQueuePresentKHR(wine_queue_from_handle(params->queue)->host_queue, &pPresentInfo_host);
+    params->pPresentInfo = &pPresentInfo_host;
+    wine_vkQueuePresentKHR(params);
     free_conversion_context(ctx);
     return STATUS_SUCCESS;
 }
@@ -47095,7 +47095,7 @@
     thunk64_vkGetShaderInfoAMD,
     thunk64_vkGetShaderModuleCreateInfoIdentifierEXT,
     thunk64_vkGetShaderModuleIdentifierEXT,
-    thunk64_vkGetSwapchainImagesKHR,
+    wine_vkGetSwapchainImagesKHR,
     thunk64_vkGetValidationCacheDataEXT,
     thunk64_vkInitializePerformanceApiINTEL,
     thunk64_vkInvalidateMappedMemoryRanges,

