--- a/dlls/winevulkan/vulkan.c
+++ b/dlls/winevulkan/vulkan.c
@@ -22,6 +22,7 @@
 #endif

 #include "config.h"
+#include <math.h>
 #include <time.h>
 #include <stdlib.h>

@@ -1470,6 +1470,7 @@
     struct wine_phys_dev *phys_dev = wine_phys_dev_from_handle(handle);
     struct wine_surface *surface = wine_surface_from_handle(surface_handle);
     VkResult res;
+    VkExtent2D user_res;

     res = phys_dev->instance->funcs.p_vkGetPhysicalDeviceSurfaceCapabilitiesKHR(phys_dev->phys_dev,

@@ -1475,6 +1475,13 @@
     if (res == VK_SUCCESS)
         adjust_max_image_count(phys_dev, capabilities);

+    if (res == VK_SUCCESS && vk_funcs->query_fs_hack &&
+            vk_funcs->query_fs_hack(surface->driver_surface, NULL, &user_res, NULL, NULL, NULL, NULL)){
+        capabilities->currentExtent = user_res;
+        capabilities->minImageExtent = user_res;
+        capabilities->maxImageExtent = user_res;
+    }
+
     return res;
 }

@@ -1490,6 +1490,7 @@
     struct wine_surface *surface = wine_surface_from_handle(surface_info->surface);
     VkPhysicalDeviceSurfaceInfo2KHR_host host_info;
     VkResult res;
+    VkExtent2D user_res;

     host_info.sType = surface_info->sType;

@@ -1501,6 +1501,13 @@
     if (res == VK_SUCCESS)
         adjust_max_image_count(phys_dev, &capabilities->surfaceCapabilities);

+    if (res == VK_SUCCESS && vk_funcs->query_fs_hack &&
+            vk_funcs->query_fs_hack(wine_surface_from_handle(surface_info->surface)->driver_surface, NULL, &user_res, NULL, NULL, NULL, NULL)){
+        capabilities->surfaceCapabilities.currentExtent = user_res;
+        capabilities->surfaceCapabilities.minImageExtent = user_res;
+        capabilities->surfaceCapabilities.maxImageExtent = user_res;
+    }
+
     return res;
 }

@@ -1767,3 +1804,6 @@
 {
     return !!vk_funcs->p_vkGetDeviceProcAddr(device->device, name);
 }
+
+
+#include "vulkan-fsr-include.c"
\ No newline at end of file
