--- a/dlls/ntdll/unix/virtual.c	2020-07-03 23:34:44.000000000 +0400
+++ b/dlls/ntdll/unix/virtual.c	2020-07-13 23:26:27.000538849 +0400
@@ -62,6 +62,7 @@
 #include "wine/rbtree.h"
 #include "unix_private.h"
 #include "wine/debug.h"
+#include "esync.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(virtual);
 WINE_DECLARE_DEBUG_CHANNEL(module);
@@ -3323,6 +3324,15 @@
  */
 void CDECL virtual_set_large_address_space(void)
 {
+  
+    printf("Begin unix esync load hack\n");
+  
+    //Hack hack hack
+    activate_esync();
+    if (do_esync())
+      esync_init();
+  
+  
     IMAGE_NT_HEADERS *nt = RtlImageNtHeader( NtCurrentTeb()->Peb->ImageBaseAddress );
 
     if (!(nt->FileHeader.Characteristics & IMAGE_FILE_LARGE_ADDRESS_AWARE)) return;
