--- sync.c	2020-08-15 01:37:58.000000000 +0400
+++ wine-wine-5.15/dlls/ntdll/unix/sync.c	2020-08-16 17:38:55.441104806 +0400
@@ -2521,10 +2521,14 @@
         return STATUS_NOT_IMPLEMENTED;
 
     val = *futex;
+    
+    RtlLeaveCriticalSection( cs );
+    status = wait_cv( futex, val, timeout );
+    RtlEnterCriticalSection( cs );
 
+    /*
     if (cs->RecursionCount == 1)
     {
-        /* FIXME: simplified version of RtlLeaveCriticalSection/RtlEnterCriticalSection to avoid imports */
         cs->RecursionCount = 0;
         cs->OwningThread   = 0;
         if (InterlockedDecrement( &cs->LockCount ) >= 0) fast_RtlpUnWaitCriticalSection( cs );
@@ -2536,6 +2540,7 @@
         cs->RecursionCount = 1;
     }
     else status = wait_cv( futex, val, timeout );
+    */
     return status;
 }
 
