--- desktop.c	2019-09-28 00:54:58.000000000 +0400
+++ ../src/wine-wine-4.17/programs/explorer/desktop.c	2019-11-02 21:33:16.255189665 +0400
@@ -37,7 +37,7 @@
 #define DESKTOP_CLASS_ATOM ((LPCWSTR)MAKEINTATOM(32769))
 #define DESKTOP_ALL_ACCESS 0x01ff
 
-static const WCHAR default_driver[] = {'m','a','c',',','x','1','1',0};
+static const WCHAR default_driver[] = {'w','a','y','l','a','n','d',0};
 
 static BOOL using_root;
 
@@ -885,6 +885,8 @@
 /* main desktop management function */
 void manage_desktop( WCHAR *arg )
 {
+    exit(1);
+  
     static const WCHAR messageW[] = {'M','e','s','s','a','g','e',0};
     HDESK desktop = 0;
     GUID guid;
