--- a/dlls/user32/driver.c	2020-03-28 21:10:34.000000000 +0400
+++ b/dlls/user32/driver.c	2020-03-28 21:10:29.000000000 +0400
@@ -60,6 +60,26 @@
     WCHAR path[MAX_PATH];
     WCHAR key[ARRAY_SIZE(key_pathW) + ARRAY_SIZE(displayW) + 40];
     UINT guid_atom;
+    
+     //TODO add defines
+    ret = (*module = LoadLibraryA( "C:\\windows\\system32\\winewayland.drv" )) != NULL;
+    if (!ret) {
+      TRACE( "failed to load 1 %d \n" );
+      ret = (*module = LoadLibraryA( "winewayland.drv" )) != NULL;
+      if (!ret) {
+        ret = (*module = LoadLibraryA( "C:\windows\system32\winewayland.drv" )) != NULL;
+        ERR( "failed to load 2 %d \n", GetLastError() );
+        if (!ret) {
+          ERR( "failed to load 3 %d \n", GetLastError() );
+        }
+      }
+    }
+    if( ret ) {
+      return ret;
+    }
+      
+    
+    
 
     USER_CheckNotLock();
 
